### PR TITLE
switch podman image scp from depending on machinectl to just os/exec

### PR DESCRIPTION
**Pull request by cdoern**
switch podman image scp from depending on machinectl to just os/exec

machinectl does not propogate error messages and adds extra lines in the output, exec.Cmd is able to clear the env besides PATH and TERM, and use the given UID and GID to execute the command properly.

machinectl is still used to create a user session. Ubuntu support is limited by this.

Signed-off-by: cdoern <cdoern@redhat.com>

### CODE REVIEW

**Review by Luap99 — CHANGES_REQUESTED** [review]
This does not work because systemd-run does not create the required login session.

I think you have to use both, first run `machinectl ... sleep inf`  to make sure that the login session is created (just start the process), then run your systemd-run command, when that is done kill the machinectl process.

**Comment by cdoern** [comment]
@Luap99 think I got all of the necessary changes in there, just made a utils function to start the machinectl sleep.

**Comment by eriksjolund** [comment]
> This does not work because systemd-run does not create the required login session.
> 
> I think you have to use both, first run `machinectl ... sleep inf` to make sure that the login session is created (just start the process), then run your systemd-run command, when that is done kill the machinectl process.

I wonder if  `machinectl ... sleep inf`  is really needed.
It seems the command
`systemd-run --machine=test11@ --quiet --user --collect --pipe --wait ...`
starts a login session.

```
[root@laptop ~]# useradd test11
[root@laptop ~]# systemd-run --machine=test11@ --quiet --user --collect --pipe --wait loginctl
SESSION  UID USER     SEAT  TTY
      2 1000 esjolund seat0 tty2
     32 1019 test11         

2 sessions listed.
[root@laptop ~]# systemd-run --machine=test11@ --quiet --user --collect --pipe --wait systemctl --user status
● laptop
    State: running
     Jobs: 0 queued
   Failed: 0 units
    Since: Sat 2022-01-15 08:10:23 CET; 131ms ago
   CGroup: /user.slice/user-1019.slice/user@1019.service
           ├─session.slice 
           │ └─dbus-broker.service 
           │   ├─6371 /usr/bin/dbus-broker-launch --scope user
           │   └─6374 dbus-broker --log 4 --controller 9 --machine-id c341780e88a04800982a41eea47639d3 --max-bytes 100000000000000 --max-fds 25000000000000 --max-matches 5000000000
           ├─app.slice 
           │ └─run-u0.service 
           │   ├─6376 systemctl --user status
           │   └─6378 less
           └─init.scope 
             ├─6346 /usr/lib/systemd/systemd --user
             └─6350 (sd-pam)
[root@laptop ~]# cat /etc/fedora-release 
Fedora release 35 (Thirty Five)
[root@laptop ~]# 
```

File: cmd/podman/images/scp.go
Comment (edsantiago) [inline cmd/podman/images/scp.go:359]: Hint: if something goes wrong here, the maintenance programmer might really appreciate seeing the actual command string that was invoked.

File: pkg/domain/infra/abi/images.go
Comment (edsantiago) [inline pkg/domain/infra/abi/images.go:878]: Could you do this as a `defer`, so a failure in lines 874-876 don't leave a stray process?

**Comment by cdoern** [comment]
> > This does not work because systemd-run does not create the required login session.
> > I think you have to use both, first run `machinectl ... sleep inf` to make sure that the login session is created (just start the process), then run your systemd-run command, when that is done kill the machinectl process.
> 
> I wonder if `machinectl ... sleep inf` is really needed. It seems the command `systemd-run --machine=test11@ --quiet --user --collect --pipe --wait ...` starts a login session.
> 
> ```
> [root@laptop ~]# useradd test11
> [root@laptop ~]# systemd-run --machine=test11@ --quiet --user --collect --pipe --wait loginctl
> SESSION  UID USER     SEAT  TTY
>       2 1000 esjolund seat0 tty2
>      32 1019 test11         
> 
> 2 sessions listed.
> [root@laptop ~]# systemd-run --machine=test11@ --quiet --user --collect --pipe --wait systemctl --user status
> ● laptop
>     State: running
>      Jobs: 0 queued
>    Failed: 0 units
>     Since: Sat 2022-01-15 08:10:23 CET; 131ms ago
>    CGroup: /user.slice/user-1019.slice/user@1019.service
>            ├─session.slice 
>            │ └─dbus-broker.service 
>            │   ├─6371 /usr/bin/dbus-broker-launch --scope user
>            │   └─6374 dbus-broker --log 4 --controller 9 --machine-id c341780e88a04800982a41eea47639d3 --max-bytes 100000000000000 --max-fds 25000000000000 --max-matches 5000000000
>            ├─app.slice 
>            │ └─run-u0.service 
>            │   ├─6376 systemctl --user status
>            │   └─6378 less
>            └─init.scope 
>              ├─6346 /usr/lib/systemd/systemd --user
>              └─6350 (sd-pam)
> [root@laptop ~]# cat /etc/fedora-release 
> Fedora release 35 (Thirty Five)
> [root@laptop ~]# 
> ```

not sure if executing on a container is what we are looking for here, this functionality is similar to that of machinectl and I am unsure if this would propagate errors the way we want to. 

--machine operates much differently than --uid if I am not mistaken, @Luap99 @edsantiago WDYT?

**Comment by edsantiago** [comment]
I'm old and crotchety, and I really, _really_ do not want to learn about systemd-run and all this newfangled crap. Can someone ELI5 why `sudo` or `su -l -c "%s" USER` aren't enough? (preferably documenting this in the code itself, so noone else has to ask again)

**Comment by Luap99** [comment]
I think --machine is what we want because it seems to create the login session. I just was never able to make it work on my test vm. You should still get the correct exit code according to the man page.

**Comment by Luap99** [comment]
> I'm old and crotchety, and I really, _really_ do not want to learn about systemd-run and all this newfangled crap. Can someone ELI5 why `sudo` or `su -l -c "%s" USER` aren't enough? (preferably documenting this in the code itself, so noone else has to ask again)

Because both sudo and su do not create a login session they just change the uid/gid. Podman needs the systemd session to function properly. The only way su/sudo will work when you set enable-linger for that user set.

**Comment by edsantiago** [comment]
>The only way su/sudo will work when you set enable-linger for that user set.

But podman is completely unusable without `linger`, so in what possible situation would anyone use `podman image scp` on a user that doesn't have the magic `linger`?

**Comment by Luap99** [comment]
You do not need linger!!!! If you login in you systemd as a user systemd will create the user session for you and you can use podman just fine. If you logout the session is destroyed so all containers are stopped. 
Linger is only needed when you want to run containers at boot or after logout.

**Comment by cdoern** [comment]
also @Luap99 I am having the same issue as you, `--machine` does not work on my machine and reports: `Failed to get machine PTY: No machine 'charliedoern@'` same fails with `charliedoern@.host`

**Comment by Luap99** [comment]
@cdoern On which os did you test this? I tested on f34 and it failed but I just tried with f35 and @eriksjolund commands worked there.

**Comment by cdoern** [comment]
@Luap99 ubuntu and f34

**Comment by eriksjolund** [comment]
I tested a few examples on Fedora 35 with a brand new user. (linger is off). It seems to work.
```
[root@laptop ~]# cat /etc/fedora-release 
Fedora release 35 (Thirty Five)
[root@laptop ~]# useradd test26
[root@laptop ~]# systemd-run --machine=test26@ --quiet --user --collect --pipe --wait podman run --rm docker.io/library/alpine echo hello
Trying to pull docker.io/library/alpine:latest...
Getting image source signatures
Copying blob 59bf1c3509f3 done  
Copying config c059bfaa84 done  
Writing manifest to image destination
Storing signatures
hello
[root@laptop ~]# systemd-run --machine=test26@ --quiet --user --collect --pipe --wait /bin/false ; echo $?
1
[root@laptop ~]# systemd-run --machine=test26@ --quiet --user --collect --pipe --wait /bin/true ; echo $?
0
[root@laptop ~]# systemd-run --machine=test26@ --quiet --user --collect --pipe --wait ls -l /non-existent ; echo $?
ls: cannot access '/non-existent': No such file or directory
2
[root@laptop ~]# systemd-run --machine=test26@ --quiet --user --collect --pipe --wait /bin/bash -c "exit 34" ; echo $?
34
[root@laptop ~]# 
```
Relevant PR 
https://github.com/systemd/systemd/pull/17967
hence __systemd v248__

**Comment by cdoern** [comment]
@eriksjolund I think we are going to stick with the current approach as a fix for the moment, but I am going to look into if there is a better way to start the login session. Your approach is preferable but does not work on f34 or ubuntu.

**Comment by edsantiago** [comment]
It does not seem to be working when run as root (on my f35 laptop):
```console
$ sudo bin/podman image scp foo.bar/nonesuch/c_zr5sx1le5n:mytag esm@localhost::
Running as unit: run-re976e3d02b7c4edea1205642220f85b8.service
Running as unit: run-r04f0570cc1cd4e778ed6368f137c8032.service
$ bin/podman images
[ does not show foo.bar/nonesuch/anything]
```
Note also that the *expected* output of `image scp` is 'Copying blob...Writing manifest...Storing signatures...' but none of that appears, only `Running as unit`. Could you test this as root?

**Comment by cdoern** [comment]
sorry @edsantiago left an extra argument in `execSystemd` works now as expected.

**Comment by edsantiago** [comment]
Now I get:
```console
$ sudo bin/podman image scp foo.bar/nonesuch/c_zr5sx1le5n:mytag esm@localhost::
Error: exit status 208
```
Adding `--log-level=debug` yields:
```
DEBU[0000] Executing via systemd: "/usr/bin/sudo /usr/bin/systemd-run --pty -q --wait --uid 0 /home/esm/src/atomic/2018-02.podman/libpod/bin/podman --log-level=debug save --output /var/tmp/podman997723794 foo.bar/nonesuch/c_zr5sx1le5n:mytag"
```

**Comment by cdoern** [comment]
@edsantiago this seems to be an selinux issue, https://bugzilla.redhat.com/show_bug.cgi?id=1559409 not sure how to get around this one...

**Comment by Luap99** [comment]
`/usr/bin/sudo /usr/bin/systemd-run --pty -q --wait --uid 0 /home/esm/src/atomic/2018-02.podman/libpod/bin/podman --log-level=debug save --output /var/tmp/podman997723794 foo.bar/nonesuch/c_zr5sx1le5n:mytag` The command looks wrong should't it be `--uid <uid from esm>` instead of `--uid 0` which is pointless since it should load the image as user.

**Comment by Luap99** [comment]
Also why do you use sudo when you are already root

**Comment by cdoern** [comment]
@Luap99 in the example @edsantiago  is giving the save is `--uid 0` and the load is `--uid <edsantiago>`

**Comment by cdoern** [comment]
--collect --pipe works on fedora rather than --pty, let me try that

**Comment by edsantiago** [comment]
/hold

Please wait until #12797 merges, then rebase on it. This PR cannot merge without at least some basic minimal testing. I'm doing it manually so far, but that's costing me all day.

**Comment by cdoern** [comment]
sgtm @edsantiago the issues here are between selinux and systemd-run so I am trying to find a way to get this to work on fedora, works well on ubuntu.

**Comment by cdoern** [comment]
@Luap99 @edsantiago systemd-run seems to be completely non functional when using --pty or --machine or --user on f35 due to selinux permissions as well as systemctl --user quirks. Going to investigate other options here because utilizing systemd-run is making this much more complicated than it has to be

cc @rhatdan in case theres an easy way to get around --pty being prohibited with selinux

**Comment by edsantiago** [comment]
@cdoern please rebase on main in order to pick up tests. Then remove the NO TESTS NEEDED, and apply the following diffs
```diff
diff --git a/test/system/120-load.bats b/test/system/120-load.bats
index f5ba93d8a..8bf785081 100644
--- a/test/system/120-load.bats
+++ b/test/system/120-load.bats
@@ -104,10 +104,6 @@ verify_iid_and_name() {
     # If we can't sudo, we can't test.
     _sudo true || skip "cannot sudo to $notme"

-    # FIXME FIXME FIXME: it'd be reeeeeeally nice if we could pass --root
-    #                    to the non-self user, hence avoid vandalizing
-    #                    their storage.
-
     # Preserve digest of original image; we will compare against it later
     run_podman image inspect --format '{{.Digest}}' $IMAGE
     src_digest=$output
@@ -117,12 +113,8 @@ verify_iid_and_name() {
     run_podman tag $IMAGE $newname

     # Copy it there.
-    # FIXME: the first '.*' in the expect string below is unfortunate; it's
-    #        a workaround for Ubuntu which gripes:
-    #        "warning.*defaulting to su since machinectl is not available"
-    #        Reexamine this once #12829 is fixed
     run_podman image scp $newname ${notme}@localhost::
-    is "$output" ".*Copying blob .*Copying config.*Writing manifest.*Storing signatures"
+    is "$output" "Copying blob .*Copying config.*Writing manifest.*Storing signatures"

     # confirm that image was copied. FIXME: also try $PODMAN image inspect?
     _sudo $PODMAN image exists $newname
@@ -132,13 +124,6 @@ verify_iid_and_name() {
     run_podman image scp -q ${notme}@localhost::$newname

     expect="Loaded image(s): $newname"
-    # FIXME FIXME FIXME: ubuntu has no machinectl, emits useless warning message instead
-    if ! is_rootless; then
-        # FIXME: root on fedora uses machinectl, which emits useless \n and \r (#12829)
-        NL=$'\n'
-        CR=$'\r'
-        expect="$NL$expect$CR"
-    fi
     is "$output" "$expect" "-q silences output"

     # Confirm that we have it, and that its digest matches our original
@@ -150,17 +135,14 @@ verify_iid_and_name() {
     run_podman untag $IMAGE $newname

     # Negative test for nonexistent image.
-    # FIXME FIXME: cannot test on root, because it uses machinectl (#12829)
-    if is_rootless; then
-        # FIXME: error message is 2 lines, the 2nd being "exit status 125".
-        # FIXME: is that fixable, or do we have to live with it?
-        nope="nope.nope/nonesuch:notag"
-        run_podman 125 image scp ${notme}@localhost::$nope
-        is "$output" "Error: $nope: image not known.*" "Pulling nonexistent image"
-
-        run_podman 125 image scp $nope ${notme}@localhost::
-        is "$output" "Error: $nope: image not known.*" "Pushing nonexistent image"
-    fi
+    # FIXME: error message is 2 lines, the 2nd being "exit status 125".
+    # FIXME: is that fixable, or do we have to live with it?
+    nope="nope.nope/nonesuch:notag"
+    run_podman 125 image scp ${notme}@localhost::$nope
+    is "$output" "Error: $nope: image not known.*" "Pulling nonexistent image"
+
+    run_podman 125 image scp $nope ${notme}@localhost::
+    is "$output" "Error: $nope: image not known.*" "Pushing nonexistent image"

     # Negative test for copying to a different name
     run_podman 125 image scp $IMAGE ${notme}@localhost::newname:newtag
```
You may want to evaluate the other FIXMEs. Whether or not you remove the `ubuntu skip` is up to you.

**Comment by Luap99** [comment]
@cdoern It still does not make sense to use systemd-run and sudo when you are already root. You can just exec the podman command normally in this case since you are already the correct user. The systemd-run part is only needed when you are root and want exec podman rootless. If I start scp rootless then only sudo is needed to start the other command as root.

**Comment by cdoern** [comment]
@Luap99 I am going to circle back to this later today, i see what you're saying about systemd-run and I only use it in one place (unnecessarily I think) I will eliminate the parts I can, apply @edsantiago tests/diffs and repush this afternoon

**Comment by cdoern** [comment]
@edsantiago system test failures here :/ can't figure out if it is an scp issue or a test issue, any ideas?

**Comment by edsantiago** [comment]
The error in both cases (f34/f35 root) is:
```
...level=error msg="reading system config \"/tmp/podman-bats-1JwO7Hc.containers.conf\": [...]: permission denied"
```
That definition of containers.conf comes from https://github.com/containers/podman/blob/094665ac0cccc04c7797881ff2a27ff330e9a72e/test/system/helpers.bash#L39-L54

Hypothesis: this new systemd-foo scheme is (incorrectly) initializing the environment of the su'ed podman.

Confirmation:
```console
# touch /root/containers-foo.conf
# chmod 0 !$
# CONTAINERS_CONF=!$ /path/to/podman image scp myimage esm@localhost::
...
ERRO[0000] finding config on system: CONTAINERS_CONF file: stat /root/containers-foo.conf: permission denied
Error: exit status 1
```
Solution: tell systemd not to transfer root environment to user?

**Comment by cdoern** [comment]
issue with that @edsantiago is systemd-run is out of the picture and we are using runuser which says that it resets all of the environmental variables to the user that is being logged into, which is what should be happening, right? https://man7.org/linux/man-pages/man1/runuser.1.html

**Comment by edsantiago** [comment]
>we are using runuser which says that it resets all of the environmental variables

My reading of `runuser(1)` is the opposite: it *does not* reset environment, unless you feed it `-l`, which (apparently) you cannot do in combination with `-u`.

**Comment by rhatdan** [comment]
BTW runuser == su, without going through PAM stack.  I wrote it in another life time.

**Comment by cdoern** [comment]
think I got it, `sudo runuser -l charliedoern -c "podman info" works, I will do that. @rhatdan yeah I am really just looking for alternatives that work on ubuntu (by default) and fedora, systemd-run didn't want to cooperate.

**Comment by cdoern** [comment]
I am unsure if `-l` creates a persistent login session because `sudo runuser -l testingUser -c loginctl` only shows charliedoern and root. I will leave the machinectl login in here until that is figured out, @Luap99 do you have any idea if runuser does the login as we need it to?

**Comment by cdoern** [comment]
@edsantiago @Luap99 thoughts on this? I am not sure if runuser creates a login session so I left the machinectl login as is.

**Comment by edsantiago** [comment]
I'm sorry, I can't help. "User session" is not a UNIX concept, it's a systemd concept, and one that I've never been able to understand. @giuseppe might know.

**Comment by cdoern** [comment]
Ok, that makes sense @rhatdan I think this is all set then @containers/podman-maintainers PTAL

File: pkg/domain/infra/abi/images.go
Comment (giuseppe) [inline pkg/domain/infra/abi/images.go:866]: I think after the `Kill()` you either need to `Wait()` for the process or `Release()` it, otherwise you leak it

File: pkg/domain/infra/abi/images.go
Comment (giuseppe) [inline pkg/domain/infra/abi/images.go:869]: completely untested, but wouldn't something like the following:

```go
	cmd := utils.CreateSCPCommand(args, []string{strings.Join(command, " ")})
	cmd.SysProcAttr = &syscall.SysProcAttr{
		Credential: &syscall.Credential{
			Uid:         execUser.Uid,
			Gid:         execUser.Gid,
			Groups:      nil,
			NoSetGroups: false,
		},
	}
```

run the command with the specified UID/GID without requiring `runuser`?

Comment (giuseppe) [reply pkg/domain/infra/abi/images.go:869]: Wasn't that leak already in `CreateSCPCommand`?

AFAICS `runuser` doesn't clean up the environment:

```
$ sudo XDG_RUNTIME_DIR=wrong runuser -u gscrivano printenv XDG_RUNTIME_DIR
wrong
```

In any case, you are right, and I think we need to drop the Environ block.

Comment (rhatdan) [reply pkg/domain/infra/abi/images.go:869]: I agree we should block the environment.

Comment (cdoern) [reply pkg/domain/infra/abi/images.go:869]: What do you mean the environment is leaked? Doesnt -l clear lol env variables except for those specified? @Luap99 `clears all the environment variables except for TERM and variables specified by --whitelist-environment`

Comment (cdoern) [reply pkg/domain/infra/abi/images.go:869]: I think -u and -l function differently according to the man pages

Comment (Luap99) [reply pkg/domain/infra/abi/images.go:869]: Yes `-l` will create a clean env, I was commenting on the suggestion from @giuseppe

Comment (cdoern) [reply pkg/domain/infra/abi/images.go:869]: ah okay, sorry I was looking at this on mobile. From what i have tried over the iterations, directly calling a n exec.Cmd via a different user using the sysprocattr basically never works and always leaves something behind from the previous session.

Comment (giuseppe) [reply pkg/domain/infra/abi/images.go:875]: can you specify some Env not nil?  From the doc:

```go
	// If Env is nil, the new process uses the current process's
	// environment.
```

Comment (cdoern) [reply pkg/domain/infra/abi/images.go:875]: hmm maybe I should try to keep $TERM like runuser does and see if that works?

Comment (rhatdan) [reply pkg/domain/infra/abi/images.go:875]: I would set $TERM and $PATH

Comment (cdoern) [reply pkg/domain/infra/abi/images.go:875]: seems to work, I am getting a storage error regarding the graph driver: `ERRO[0000] User-selected graph driver "vfs" overwritten by graph driver "overlay" from database - delete libpod local files to resolve` but the command finishes successfully. This could be my machine so I will push here and let the tests run

**Comment by cdoern** [comment]
@giuseppe any idea why this is failing? This error has happened to me before in my initial iterations of scp, do I need to UID/GID map as well in the SysProcAttr?

**Comment by giuseppe** [comment]
> @giuseppe any idea why this is failing? This error has happened to me before in my initial iterations of scp, do I need to UID/GID map as well in the SysProcAttr?

no we don't neet to specify the map, it has simply to run with the specified UID/GID

File: pkg/domain/infra/abi/images.go
Comment (giuseppe) [inline pkg/domain/infra/abi/images.go:857]: would this work without specifying `TERM=`?  I think it should be `"TERM=" + term`

**Comment by giuseppe** [comment]
the tests are failing because it doesn't find the `newuidmap` executable.  Please propagate also `PATH` in addition to `TERM`

**Comment by cdoern** [comment]
I initially tried to do this with PATH as well and got the same error, could be because I was not doing `PATH=...` I will try that

**Comment by rhatdan** [comment]
/approve
/lgtm

I am going to merge, since I believe this is an improvement over what we currently have, and we should get it into podman 4.0.
@cdoern continue working on improving it. and fixing other bugs in podman image scp.

**Pull request merged**
